custom/icds_reports/views.py

```diff
@@ -1718,3 +1721,63 @@ def get(self, request, *args, **kwargs):
             return export_response(sync.get_file_from_blobdb(), 'csv', zip_name)
         except NotFound:
             raise Http404
+
+
+class DishaAPIView(View):
+
+    def message(self, message_name):
+        state_names = ", ".join(self.valid_state_names)
+        error_messages = {
+            "missing_date": "Please specify valid month and year",
+            "invalid_month": "Please specify a month that's older than a month and 5 days",
+            "invalid_state": "Please specify one of {} as state_name".format(state_names),
+            "data_is_unavailable": "No data is available for this month"
+        }
+        return {"message": error_messages[message_name]}
+
+    @method_decorator(api_auth)
+    @json_error
+    def get(self, request, *args, **kwargs):
+        # Todo; add disha API permission
+        try:
+            month = int(request.GET.get('month'))
+            year = int(request.GET.get('year'))
+        except (ValueError, TypeError):
+            return JsonResponse(self.messages('missing_date'), status=400)
+
+        # Can return only one month old data if today is after 5th, otherwise
+        #   can return two month's old data
+        query_month = date(year, month, 1)
+        today = date.today()
+        current_month = today - relativedelta(months=1) if today.day <= 5 else today
+        if query_month > current_month:
+            return JsonResponse(self.messages('invalid_month'), status=400)
+
+        state_name = self.request.GET.get('state_name')
```

Comment:
I glanced over the spec and it seems like this API is largely defined by us. If that's true, we should paginate this page in some way. Some quick calculations say that this API is going to return >20 MB responses at some point, and combined with the joins in the view it will be a heavy request that could potentially time out.

100k AWC * 40 indicators * 4 bytes each = 16 MB. There are a few states with more than 100k AWC and we would also need to add district, blocks and supervisors to this calculation.

Alternatively, we could precompute this for every state and serve those responses as static files

Reply:
Yeah, this is marked as todo above. Planning to precompute and save it. 

Re. whether we can paginate, the team that consumes this API wants a dump of data that they can use to build a dashboard. So, we can't paginate them.

Reply:
> Re. whether we can paginate, the team that consumes this API wants a dump of data that they can use to build a dashboard. So, we can't paginate them.

That doesn't make sense. We can paginate it, they can download it to their own database in chunks and append it to their own database. Unless they are going to try to serve this URL directly which will have it's own set of problems